src/tracers/tracers.cpp

```diff
@@ -0,0 +1,294 @@
+
+  if (mhd) {
+    physics->AddSwarmValue("B_x", swarm_name, real_swarmvalue_metadata);
+    physics->AddSwarmValue("B_y", swarm_name, real_swarmvalue_metadata);
+    physics->AddSwarmValue("B_z", swarm_name, real_swarmvalue_metadata);
+  }
+
+  physics->PostFillDerivedBlock = FillTracers;
+
+  return physics;
+} // Initialize
+
+TaskStatus AdvectTracers(MeshBlockData<Real> *rc, const Real dt) {
+  using namespace LCInterp;
+  namespace p = fluid_prim;
+
+  auto *pmb = rc->GetParentPointer().get();
+  auto &sc = pmb->swarm_data.Get();
+  auto &swarm = sc->Get("tracers");
+
+  const auto ndim = pmb->pmy_mesh->ndim;
+
+  auto &x = swarm->Get<Real>("x").Get();
+  auto &y = swarm->Get<Real>("y").Get();
+  auto &z = swarm->Get<Real>("z").Get();
+
+  auto swarm_d = swarm->GetDeviceContext();
+
+  const std::vector<std::string> vars = {p::velocity};
+
+  PackIndexMap imap;
+  auto pack = rc->PackVariables(vars, imap);
+
+  const int pvel_lo = imap[p::velocity].first;
+  const int pvel_hi = imap[p::velocity].second;
+
+  auto geom = Geometry::GetCoordinateSystem(rc);
+
+  // update loop.
+  const int max_active_index = swarm->GetMaxActiveIndex();
+  pmb->par_for(
+      "Advect Tracers", 0, max_active_index, KOKKOS_LAMBDA(const int n) {
+        if (swarm_d.IsActive(n)) {
+          int k, j, i;
+          swarm_d.Xtoijk(x(n), y(n), z(n), i, j, k);
+
+          // geom quantities
+          Real gcov4[4][4];
+          geom.SpacetimeMetric(0.0, x(n), y(n), z(n), gcov4);
+          Real lapse = geom.Lapse(0.0, x(n), y(n), z(n));
+          Real shift[3];
+          geom.ContravariantShift(0.0, x(n), y(n), z(n), shift);
+
+          // Get shift, W, lapse
+          const Real Wvel_X1 = LCInterp::Do(0, x(n), y(n), z(n), pack, pvel_lo);
+          const Real Wvel_X2 = LCInterp::Do(0, x(n), y(n), z(n), pack, pvel_lo + 1);
+          const Real Wvel_X3 = LCInterp::Do(0, x(n), y(n), z(n), pack, pvel_hi);
+          const Real Wvel[] = {Wvel_X1, Wvel_X2, Wvel_X3};
+          const Real W = phoebus::GetLorentzFactor(Wvel, gcov4);
+          const Real vel_X1 = Wvel_X1 / W;
+          const Real vel_X2 = Wvel_X2 / W;
+          const Real vel_X3 = Wvel_X3 / W;
+
+          const Real rhs1 = (lapse * vel_X1 - shift[0]) * dt;
+          const Real rhs2 = (lapse * vel_X2 - shift[1]) * dt;
+          Real rhs3 = 0.0;
+          if (ndim > 2) {
+            rhs3 = (lapse * vel_X3 - shift[2]) * dt;
+          }
+
+          x(n) += rhs1;
+          y(n) += rhs2;
+          z(n) += rhs3;
+
+          bool on_current_mesh_block = true;
+          swarm_d.GetNeighborBlockIndex(n, x(n), y(n), z(n), on_current_mesh_block);
+        }
+      });
+
+  return TaskStatus::complete;
+} // AdvectTracers
+
+/**
+ * FillDerived function for tracers.
+ * Registered Quantities (in addition to t, x, y, z):
+ * rho, T, ye, vel, W_lorentz, pressure, lapse, shift, entropy, detgamma, B
+ **/
+void FillTracers(MeshBlockData<Real> *rc) {
```

Comment:
I was initially a little confused why `FillTracers` and `AdvectTracers` aren't a single kernel---but I think it makes sense to split things this way... In fact, in the long term, we should consider moving `FillTracers` from `FillDerivedBlock` to `UserWorkBeforeOutput`. That way this interpolation only happens on I/O, not every step. That's for future us, though. :)

Reply:
This is now done, right? :)